### PR TITLE
cmake: version.cmake: enforce single byte limits

### PR DESCRIPTION
**Pull request by JordanYates**
Enforce the single byte size limits on the numeric fields in `VERSION`.

The `VERSION` numeric fields are used to construct the `uint32_t` `KERNELVERSION` and `APPVERSION` values, which only allocates a single byte for each field.

Additionally adds a test application for the version file generation.

### CODE REVIEW

**Review by nordicjm — CHANGES_REQUESTED** [review]
Then change the regex so it is limited to 1 byte for tweak? The MCUboot fields have to fit in with the zephyr fields, this was decided when the feature was introduced in the architecture working group, so whilst MCUboot has a 32-bit value for build number, it would be restricted to 8-bit when using the default zephyr VERSION file, however, a user is free to use their own way of setting build version by setting the MCUboot signing version Kconfig to their choosing.

**Comment by JordanYates** [comment]
> Then change the regex so it is limited to 1 byte for tweak? The MCUboot fields have to fit in with the zephyr fields, this was decided when the feature was introduced in the architecture working group, so whilst MCUboot has a 32-bit value for build number, it would be restricted to 8-bit when using the default zephyr VERSION file, however, a user is free to use their own way of setting build version by setting the MCUboot signing version Kconfig to their choosing.

What logic would you suggest for performing that limiting? Build error if over, or silently take the lowest 8 bits?
I don't suppose you remember at which point this was discussed, I would be interested in reading any comments. Although the code that consumes the constructed value (`kernel_version.h`) was introduced at least 9 years ago...

Note that doing any limiting here will make #73380 impossible to merge, as it is setting the same value, and 8 bits is not enough to be reasonably unique in that context.

**Comment by nordicjm** [comment]
> What logic would you suggest for performing that limiting? Build error if over, or silently take the lowest 8 bits?

cmake error since the version file can be adjusted and configuration can re-run. Or if not that then a warning and truncate.

>  I don't suppose you remember at which point this was discussed, I would be interested in reading any comments. Although the code that consumes the constructed value (kernel_version.h) was introduced at least 9 years ago...

It is in the comments here: https://github.com/zephyrproject-rtos/zephyr/issues/56605

**Comment by JordanYates** [comment]
Updated the PR to document the value limits as they currently exist and enforce said limits in cmake.

File: tests/lib/app_version/CMakeLists.txt
Comment (nordicjm) [inline tests/lib/app_version/CMakeLists.txt:3]: `tests/cmake` might be a better location for it (optional)

File: cmake/modules/version.cmake
Comment (nordicjm) [inline cmake/modules/version.cmake:22]: should be 7 not 07, that would actually treat it as octal...

File: tests/lib/app_version/CMakeLists.txt
Comment (JordanYates) [reply tests/lib/app_version/CMakeLists.txt]: done

File: cmake/modules/version.cmake
Comment (nordicjm) [reply cmake/modules/version.cmake:22]: also text here needs updating

Comment (JordanYates) [reply cmake/modules/version.cmake]: Replaced the entire block with the numbers from `version.h`, order and all.

Comment (JordanYates) [reply cmake/modules/version.cmake]: which is not what is should be...

Comment (JordanYates) [reply cmake/modules/version.cmake]: The text there is fine, as `${type}_VERSION_TWEAK` is the output variable from the cmake module, even if the generated C define has a different name format.

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Review by tejlmand — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**
